sno/diff.py

```diff
@@ -26,360 +25,241 @@ class Conflict(Exception):
-
-        return True
-
-    def dataset_counts(self, dataset):
-        """Returns a dict containing the count of each type of diff, for a particular dataset."""
-        return {k: len(v) for k, v in self._data[dataset.path].items()}
-
-    def counts(self):
-        """
-        Returns multiple dataset_counts dicts, one for each dataset touched by this diff.
-        The dataset_counts dicts are returned in a top-level dict keyed by dataset path.
-        """
+    def __len__(self):
+        return len(self.children)
 
-        return {
-            dataset.path: self.dataset_counts(dataset) for dataset in self.datasets()
-        }
+    def __bool__(self):
+        return any(bool(child) for child in self.children.values())
+
+    def __str__(self):
+        return f"Diff({self.self_key}, children={{{','.join(str(k) for k in self.children.keys())}}}))"
+
+    __repr__ = __str__
 
-    def __repr__(self):
-        return repr(self._data)
+    def __invert__(self):
+        return Diff(self.self_key, (~c for c in self.children.values()))
 
-    def datasets(self):
-        return self._datasets.values()
+    def __add__(self, other):
+        # FIXME: this algorithm isn't perfect when renames are involved.
+
+        result = []
+        for child_key in self.children.keys() | other.children.keys():
+            lhs = self.children.get(child_key)
+            rhs = other.children.get(child_key)
+            if lhs and rhs:
+                both = lhs + rhs
+                if both:
+                    result.append(both)
+            else:
+                result.append(lhs or rhs)
+        return Diff(self.self_key, result)
+
+    def __iadd__(self, other):
+        self.children = (self + other).children
+        return self
 
     def to_filter(self):
         """
-        Returns a filter object - see filter_util.py - that matches all features affected by this diff.
+        Returns the set of all keys of all the Deltas that are children of this Diff.
+        If the children are Diffs, not Deltas, returns instead dict of {child.key: child.to_filter()}
         """
-        return {dataset.path: self._dataset_pks(dataset) for dataset in self.datasets()}
-
-    def _dataset_pks(self, dataset):
-        """Returns the set of all pks affected by this diff within a particular dataset."""
-        ds_data = self._data[dataset.path]
-        primary_key = dataset.primary_key
-        inserts = [str(o[primary_key]) for o in ds_data["I"]]
-        deletes = [str(o[primary_key]) for o in ds_data["D"].values()]
-        update_olds = [str(o[0][primary_key]) for o in ds_data["U"].values()]
-        update_news = [str(o[1][primary_key]) for o in ds_data["U"].values()]
-        return set(inserts + deletes + update_olds + update_news)
+        try:
+            # If children are Diffs: recursively call to_filter() on them too.
+            return {child.key: child.to_filter() for child in self.children.values()}
+        except AttributeError:
```

Comment:
Diffs can only contain _either_ Diffs or Deltas. It might be worth enforcing that when you add a child, and storing what type of diff this is. Then this AttributeError stuff could be avoided

Reply:
Done